### PR TITLE
Feature/Suggest Widget

### LINKED ISSUE #4960: Get suggested widgets along with execute API or query response

### PR DESCRIPTION
**Pull request by AnaghHegde**
## Description

As part of editor simplification and new navigation, this change set address the widget suggestion feature. Based on some basic rules(which will look for specific fields and records count) the suggested widget is sent along with executed query response. 

Fixes 4960 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Unit Tests 
- Manual testing

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:white_circle: Total coverage has not changed</summary>
No changes to code coverage between the base branch and the head branch</details>

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/NewActionServiceImpl.java [hunk 1]
```diff
@@ -693,6 +698,31 @@ private ActionExecutionResult addDataTypes(ActionExecutionResult result) {
         return result;
     }
 
+    /**
+     * Suggest the best widget to the query response. We currently planning to support List, Select, Table and Chart widgets
+     */
+    private String getSuggestedWidget(Object data) {
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/NewActionServiceImpl.java:704]: Can we add some tests for calling this function directly with various crazy combinations of widgets and of object types? I can see this function is going to progressively grow in complexity as we add intelligence for more widget types and it'd be great if we started off on testing this early on.

**Comment by sharat87** [comment]
@AnaghHegde, I don't know if this is in scope for this but wanted to put it out here in case we want to do this.

I see that the suggested widget field is declared to be just a `String`, right? Could we not suggest multiple widget types in the future? Like, for an array of objects with `width` and `height` of people, both a table widget and a chart widget could be good suggestions right? In that sense, should we make this a `List<String>` instead?

Taking that thought a bit further, what if we could identify that the response data has a `users` array, for which a table widget is ideal, and it also has a `history` array, for which a chart widget is ideal, for example. Then we could tell the frontend with a map like `{".users": "TABLE_WIDGET", ".history": "CHART_WIDGET"}`, so the UI can show two suggested widgets for different pieces of data in the response. If we want to go this way, we'll want a `Map<String, String>` instead.

Its possible I'm thinking too much here. @Nikhil-Nandagopal, do the above sound too farfetched? 😄

For today's case though, the implementation looks good. Lets add a few tests and we should be ready to merge.

**Comment by Nikhil-Nandagopal** [comment]
> @AnaghHegde, I don't know if this is in scope for this but wanted to put it out here in case we want to do this.
> 
> I see that the suggested widget field is declared to be just a `String`, right? Could we not suggest multiple widget types in the future? Like, for an array of objects with `width` and `height` of people, both a table widget and a chart widget could be good suggestions right? In that sense, should we make this a `List<String>` instead?
> 
> Taking that thought a bit further, what if we could identify that the response data has a `users` array, for which a table widget is ideal, and it also has a `history` array, for which a chart widget is ideal, for example. Then we could tell the frontend with a map like `{".users": "TABLE_WIDGET", ".history": "CHART_WIDGET"}`, so the UI can show two suggested widgets for different pieces of data in the response. If we want to go this way, we'll want a `Map<String, String>` instead.
> 
> Its possible I'm thinking too much here. @Nikhil-Nandagopal, do the above sound too farfetched? 😄
> 
> For today's case though, the implementation looks good. Lets add a few tests and we should be ready to merge.

Not at all. Infact the basic requirement is to suggest 1 table, 1 list, 1 select, 1 chart and 1 text if possible

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/NewActionServiceImpl.java (hunk 1 above)
Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/NewActionServiceImpl.java:704]: Yes, i have added an extra check for widget suggested in existing unit tests. I am gonna write a separate unit test for this method.

**Pull request closed without merging**